pkg/engine/renderer/dependencies.go

```diff
@@ -0,0 +1,245 @@
+			if err != nil {
+				return fmt.Errorf("error calculating hash for %s of type %s: %v", name, depType, err)
+			}
+			_, _ = depHash.Write(hash[:])
+		}
+	}
+
+	hashStr := fmt.Sprintf("%x", depHash.Sum([]byte{}))
+
+	return setTemplateHash(obj, hashStr)
+}
+
+func (de *dependencyCalculator) getHashForDependency(name string, t reflect.Type) (hashBytes, error) {
+	cache := de.cache[t]
+
+	if hash, ok := cache[name]; ok {
+		return hash, nil
+	}
+
+	dep, err := de.resourceDependency(name, t)
+	if err != nil {
+		return hashBytes{}, fmt.Errorf("failed to get dependeny %s/%s: %v", de.namespace, name, err)
+	}
+	if _, ok := dep.GetAnnotations()[kudo.SkipHashCalculationAnnotation]; ok {
+		cache[name] = hashBytes{}
+	} else {
+		yamlStr, err := calculateResourceHash(dep)
+		if err != nil {
+			return hashBytes{}, fmt.Errorf("failed to serialize dependeny %s/%s: %v", de.namespace, name, err)
+		}
+		cache[name] = md5.Sum([]byte(yamlStr)) //nolint:gosec
+	}
+
+	return cache[name], nil
+}
+
+// Calculates a stable hash from a resource
+func calculateResourceHash(obj metav1.Object) (string, error) {
+	// Namespace is ignored mostly to allow easier integration tests (which use random namespaces)
+	ns := obj.GetNamespace()
+
+	// OwnerReferences need to be skipped as they contain a changing UID
+	or := obj.GetOwnerReferences()
+
+	// Annotations are notorious for containing quickly changing strings: plan/phase/task names, uids, dates, etc.
+	ann := obj.GetAnnotations()
+
+	obj.SetNamespace("")
+	obj.SetOwnerReferences([]metav1.OwnerReference{})
+	obj.SetAnnotations(map[string]string{})
+	yamlStr, err := ToYaml(obj)
+	obj.SetNamespace(ns)
+	obj.SetOwnerReferences(or)
+	obj.SetAnnotations(ann)
+	return yamlStr, err
+}
+
+// resourceDependency returns the resource of type t with the given namespace/name, either from the passed in list of objects or the last applied configuration from the API server
+func (de *dependencyCalculator) resourceDependency(name string, t reflect.Type) (metav1.Object, error) {
+
+	// First try to find the dependency in the local list, if it's deployed in the same task we'll find it here
+	for _, obj := range de.taskObjects {
+		if reflect.TypeOf(obj).Elem() == t {
+			obj, _ := obj.(metav1.Object)
+			if obj.GetName() == name {
+				return obj, nil
+			}
+		}
+	}
+
+	// We haven't found it, so we need to query the api server to get the current version
+	dep, _ := reflect.New(t).Interface().(metav1.Object)
+	key := client.ObjectKey{
+		Namespace: de.namespace,
+		Name:      name,
+	}
+
+	err := de.Client.Get(context.TODO(), key, dep.(runtime.Object))
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve object %s/%s: %v", de.namespace, name, err)
+	}
+
+	// We don't want the hash from the object itself, because of added metadata from the api-server
+	// We use the LastAppliedConfigAnnotation that stores exactly what we applied last time
+	lastConfiguration, ok := dep.GetAnnotations()[kudo.LastAppliedConfigAnnotation]
```

Comment:
Kind of. Users editing KUDO-deployed resources behind KUDOs back is already a problem. It might work ok, if they don't modify anything that conflicts with the next KUDO Apply, and in that case it will work here as well.

This might turn out a bit tricky, but it should be less problematic than the TaskApply handling.